### PR TITLE
Feature/ch9409/statechange after finish update

### PR DESCRIPTION
**Pull request by oed**

### CODE REVIEW

File: src/muport.js [hunk 1]
```diff
@@ -112,21 +112,27 @@ class MuPort {
...
     const signedTx = this.keyring.signManagementTx(txParams)
 
     return {
       address,
       costInEther,
       finishUpdate: async () => {
-        await this.ethUtils.sendRawTx(signedTx)
+        const txHash = await this.ethUtils.sendRawTx(signedTx)
+        await this.ethUtils.waitForTx(txHash)
+        this.document = newDocument
```
Comment (christianlundkvist) [inline src/muport.js]: Ok so now you can cancel at any point before calling `finishUpdate`, which is good. What happens if the above transaction fails? Should we ideally be checking the transaction receipt and only updating the document/hash if the transaction is successful?

File: src/muport.js [hunk 2]
```diff
@@ -112,21 +112,27 @@ class MuPort {
...
-    this.document.recoveryNetwork = await this.keyring.createShares(delegateDids, didsPublicKeys)
+    // seems like the best way to do a deep clone
+    const newDocument = JSON.parse(JSON.stringify(this.document))
+    newDocument.recoveryNetwork = await this.keyring.createShares(delegateDids, didsPublicKeys)
     // save guardians
-    this.document.symEncryptedData.symEncDids = delegateDids.map((did) => this.keyring.symEncrypt(didToBuffer(did)))
-    this.documentHash = await ipfs.addJSONAsync(this.document)
+    newDocument.symEncryptedData = newDocument.symEncryptedData || {}
+    newDocument.symEncryptedData.symEncDids = delegateDids.map((did) => this.keyring.symEncrypt(didToBuffer(did)))
+    let newDocumentHash = await ipfs.addJSONAsync(newDocument)
```
Comment (christianlundkvist) [inline src/muport.js]: Ideally we could compute the hash locally without adding to IPFS here, but that can be an optimization later on.

File: src/muport.js [hunk 3]
```diff
@@ -185,17 +191,17 @@ class MuPort {
...
 
-      symEncryptedDelegateDids = delegateDids.map((did) => keyring.symEncrypt(didToBuffer(did)))
-
+      let symEncryptedDelegateDids = delegateDids.map((did) => keyring.symEncrypt(didToBuffer(did)))
+      symEncryptedData = {symEncDids: symEncryptedDelegateDids}
     }
     const publicKeys = keyring.getPublicKeys()
 
-    const doc = createMuportDocument(publicKeys, recoveryNetwork, publicProfile, {symEncDids: symEncryptedDelegateDids})
+    const doc = createMuportDocument(publicKeys, recoveryNetwork, publicProfile, symEncryptedData)
```
Comment (christianlundkvist) [inline src/muport.js]: I like this better, thanks!

File: lib/ethereum-utils.js [hunk 4]
```diff
@@ -25,7 +25,7 @@ function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { de
 var coder = require('web3/lib/solidity/coder');
 var CryptoJS = require('crypto-js');
 var bs58 = require('bs58');
-var Web3 = require('web3');
+var ethers = require('ethers');
```
[the remark this answers is not in the document]
Comment (oed) [reply lib/ethereum-utils.js]: Me too, so smooth. If you want you can have a look at replacing the web3 coder at some point, so we can remove web3 as a dep :)

File: src/muport.js (hunk 1 above)
Comment (oed) [reply src/muport.js]: Good point. I guess we can just put the `waitForTx` call and everything after it in a `try` clause?

File: src/muport.js (hunk 2 above)
Comment (oed) [reply src/muport.js]: Yeah, Ideally we'd use the real ipfs lib also :)

File: src/muport.js (hunk 1 above)
Comment (christianlundkvist) [reply src/muport.js]: Hmm not sure, I think we can start by just worrying about those.

**Pull request merged**